### PR TITLE
Add Fast cluster script

### PR DESCRIPTION
**Pull request by iwatobipen**
#### What does this implement/fix? Explain your changes.
Dear developers, I added Fast clustering script for large amount of molecules.
This script needs bayon at first, but it works very fast.

#### Any other comments?
About 200K compounds are clustered within 10 minutes. < Core i7, 16GB RAM, ubuntu 16.04 >

### CODE REVIEW

File: Contrib/Fastcluster/fastcluster.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    parsefile = open( output.split(".")[0]+"_parse.tsv", "w" )
+    inputf = open( output, "r" )
+    for line in inputf:
+        line = line.rstrip().split( "\t" )
+        cluster_id = line[0]
+        for i in range( 1, len( line )-1, 2 ) :
+            molid = line[ i ]
+            point = line[ i + 1 ]
+            parsefile.write("{}\t{}\tCLS_ID_{}\n".format( molid, point, cluster_id ))
+    parsefile.close()
```
Comment (gedeck) [inline Contrib/Fastcluster/fastcluster.py]: Use with-statement:
```
with open( output, "r" ) as inputf:
  with open( output.split(".")[0]+"_parse.tsv", "w" ) as parsefile:
    for line in inputf:
      line = line.rstrip().split( "\t" )
      cluster_id = line[0]
      for i in range(1, len(line) - 1, 2) :
        molid = line[i]
        point = line[i + 1]
        parsefile.write("{}\t{}\tCLS_ID_{}\n".format(molid, point, cluster_id))
```
There are other ways to iterate over elements of an array in pairs: https://stackoverflow.com/questions/5389507/iterating-over-every-two-elements-in-a-list

```
def grouped(iterable, n):
    "s -> (s0,s1,s2,...sn-1), (sn,sn+1,sn+2,...s2n-1), (s2n,s2n+1,s2n+2,...s3n-1), ..."
    return zip(*[iter(iterable)]*n)

with open( output, "r" ) as inputf:
  with open( output.split(".")[0]+"_parse.tsv", "w" ) as parsefile:
    for line in inputf:
      line = line.rstrip().split( "\t" )
      cluster_id = line.pop(0)
      for molid, point in grouped(line, 2):
        parsefile.write("{}\t{}\tCLS_ID_{}\n".format(molid, point, cluster_id))
```
Works in Python 3 only.

File: Contrib/Fastcluster/fastcluster.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    inputf = open( args.input, "r" )
+    n = args.n
+    c = args.c
+    output = args.output
+    tempf = open( "fp.tsv", "w" )
+    for line in inputf:
+        line = line.rstrip().split( "\t" )
+        tempf.write( smi2fp( line[0], line[1] ))
+    tempf.close()
+    res = subprocess.call( "time bayon -p -c {} -n {} fp.tsv > {}".format( c, n, output ), shell=True )
```
Comment (gedeck) [inline Contrib/Fastcluster/fastcluster.py]: does it make any sense to continue the program is `res!=0`? If not, use:
```
if res != 0:
  parser.exit("Error running bayon")
```

File: Contrib/Fastcluster/fastcluster.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            molid = line[ i ]
+            point = line[ i + 1 ]
+            parsefile.write("{}\t{}\tCLS_ID_{}\n".format( molid, point, cluster_id ))
+    parsefile.close()
+
+
+    if res == 0:
+        print( "Done!" )
+    else:
+        print( "Error :-(" )
```
Comment (gedeck) [inline Contrib/Fastcluster/fastcluster.py]: I would use the `parser.error` or `parser.exit` method. This makes sure that the exit status is set.
```
if res == 0:
  print("Done!")  # or parser.exit(message="Done!")
else:
  parser.error("Error running bayon")
```
As mentioned above, handling the case `res>0` is better done immediately after the subprocess call.

File: Contrib/Fastcluster/sdf2smi.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import argparse
+from rdkit import Chem
+sdf = Chem.SDMolSupplier("cdk2.sdf")
+f = open("cdk2.smi","w")
+for mol in sdf:
+    name = mol.GetProp("_Name")
+    smi = Chem.MolToSmiles( mol )
+    f.write( "{}\t{}\n".format(name,smi))
+f.close()
```
Comment (gedeck) [inline Contrib/Fastcluster/sdf2smi.py]: Use with statement for `f`

**Comment by iwatobipen** [comment]
Thank you for your comments. I accept all your comments and will revise my script.

File: Contrib/Fastcluster/fastcluster.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+parser.add_argument( "--output", default = "clustered.tsv" )
+parser.add_argument( "-c", help = "filename of centroid", default = "centroid.tsv" )
+
+def smi2fp( molid, smiles ):
+    mol = Chem.MolFromSmiles( smiles )
+    onbits = AllChem.GetMorganFingerprintAsBitVect( mol, 2 ).GetOnBits()
+    row = molid
+    for bit in onbits:
+        row += "\tFP_{}\t1.0".format( bit )
+    row += "\n"
```
[the remark this answers is not in the document]
Comment (iwatobipen) [reply Contrib/Fastcluster/fastcluster.py]: Thank you for your suggestion!

**Comment by iwatobipen** [comment]
reflected on Gedeck suggestion

File: Contrib/Fastcluster/fastcluster.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    parsefile = open(args.output.split(".")[0]+"_parse.tsv", "w")
+    inputf = open(args.output, "r")
+    for line in inputf:
+        line = line.rstrip().split("\t")
+        cluster_id = line[0]
+        for i in range(1, len(line)-1, 2) :
+            molid = line[ i ]
+            point = line[ i + 1 ]
+            parsefile.write("{}\t{}\tCLS_ID_{}\n".format(molid, point, cluster_id))
+    parsefile.close()
```
Comment (gedeck) [inline Contrib/Fastcluster/fastcluster.py]: You can also convert this to using the with statement. It's not really essential in a script as files will be closed at the end. However, I think it's a good idea to get into the habit of using it.

File: Contrib/Fastcluster/fastcluster.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        row += "FP_{}\t1.0".format(bit)
+    row += "\n" 
+    return row 
+
+
+if __name__ == "__main__":
+    parser = getArgParser()
+    args = parser.parse_args()
+    with open(args.input, "r") as inputf:
+        with open("fp.tsv", "w") as tempf:
```
Comment (gedeck) [inline Contrib/Fastcluster/fastcluster.py]: If you decide to only support Python 3, these two lines could also be written as:
```
 with open(args.input, "r") as inputf, open("fp.tsv", "w") as tempf:
```

File: Contrib/Fastcluster/.gitignore [hunk 8]
```diff
@@ -0,0 +1,3 @@
+*.pyc
```
Comment (greglandrum) [inline Contrib/Fastcluster/.gitignore]: I don't think this .gitignore file needs to be (or really should be) part of the PR

File: Contrib/Fastcluster/testdata/sdf2smi.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import argparse
```
Comment (greglandrum) [inline Contrib/Fastcluster/testdata/sdf2smi.py]: You could do the work that this file does a tiny bit more simply using an RDKit `SmilesMolWriter`, but doing it with Python is fine too.

File: Contrib/Fastcluster/.gitignore (hunk 8 above)
Comment (iwatobipen) [reply Contrib/Fastcluster/.gitignore]: Dear Greg,
Thank you for your advice. I removed .gitignore file.

File: Contrib/Fastcluster/testdata/sdf2smi.py (hunk 9 above)
Comment (iwatobipen) [reply Contrib/Fastcluster/testdata/sdf2smi.py]: I tried to change sdf2smi.py by using SmilesMolWriter. Thank you for your comment.

**Comment by greglandrum** [comment]
Thanks for the contribution @iwatobipen !

**Comment by iwatobipen** [comment]
I appreciate your kind advice!

**Pull request merged**